File: Algorithms/1. Sorting/Merge Sort/python/merge_sort.py
# nlog(n)
def merge_sort(nums: list[int]) -> list[int]:
    if len(nums) <= 1: return nums
    mid = len(nums) // 2
    
    left = merge_sort(nums[0:mid])
    rite = merge_sort(nums[mid:len(nums)])

    return merge(left, rite)

# helper function
def merge(left: list[int], rite: list[int]) -> list[int]:
    result = []
    l = 0
    r = 0

    while l < len(left) and r < len(rite):
        if left[l] <= rite[r]:
            result.append(left[l])
            l += 1
        else:
            result.append(rite[r])
            r += 1
    
    for k in range(l, len(left)):
        result.append(left[k])

    for k in range(r, len(rite)):
        result.append(rite[k])

    return result


print(merge_sort([]))
print(merge_sort([1]))
print(merge_sort([1, 2, 3, 4]))
print(merge_sort([4, 3, 2, 1]))
print(merge_sort([1, 2, 3, 4, 5]))
print(merge_sort([5, 4, 3, 2, 1]))
print(merge_sort([5, 4, 3, 2, 1, 9, 8, 7, 6, 0]))

File: Algorithms/1. Sorting/Merge Sort/cpp/merge_sort.cpp
#include "merge_sort.hpp"

// nlog(n)
std::vector<int> mergeSort(std::vector<int> array) {
  if (array.size() <= 1) return array;

  int mid_index = (int) array.size() / 2;
  std::vector<int> left;
  std::vector<int> rite;

  if (array.size() & 1) {
    left = std::vector<int>(array.begin(), array.end() - mid_index);
    rite = std::vector<int>(array.begin() + mid_index + 1, array.end());
  } else {
    left = std::vector<int>(array.begin(), array.end() - mid_index);
    rite = std::vector<int>(array.begin() + mid_index, array.end());
  }

  left = mergeSort(left);
  rite = mergeSort(rite);
  return merge(left, rite);
}

// helper function
std::vector<int> merge(std::vector<int> left, std::vector<int> rite) {
  std::vector<int> result;

  while (!left.empty() && !rite.empty()) {

    if (left[0] < rite[0]) {
      result.push_back(left[0]);
      left.erase(left.begin());

    } else if (left[0] > rite[0]) {
      result.push_back(rite[0]);
      rite.erase(rite.begin());

    } else {
      result.push_back(left[0]);
      left.erase(left.begin());
    }
  }

  result.insert(result.end(), left.begin(), left.end());
  result.insert(result.end(), rite.begin(), rite.end());
  return result;
}

// debugger
void print(std::vector<int> array) {
  std::cout << "{ ";
  for (int index = 0; index < array.size(); index++) {
    std::cout << array[index];
    if (index < array.size() - 1) std::cout << ", ";
  }
  std::cout << " }" << std::endl;
}
